src/inc/quic_platform_posix.h

```diff
@@ -968,73 +1046,148 @@ void
-    if (io_sqe == NULL) return FALSE; // OOM
+    BOOLEAN Enqueued = FALSE;
+    CxPlatLockAcquire(&Queue->Lock);
+    struct io_uring_sqe *io_sqe = io_uring_get_sqe(&Queue->Ring);
+    if (io_sqe == NULL) {
+        goto Exit; // OOM
+    }
     io_uring_prep_nop(io_sqe);
-    io_uring_sqe_set_data(io_sqe, sqe);
-    io_uring_submit(queue); // TODO - Extract to separate function?
-    return TRUE;
+    io_uring_sqe_set_data(io_sqe, Sqe);
+    io_uring_submit(&Queue->Ring); // TODO - Extract to separate function?
+    Enqueued = TRUE;
+Exit:
+    CxPlatLockRelease(&Queue->Lock);
+    return Enqueued;
 }
 
 inline
 uint32_t
 CxPlatEventQDequeue(
-    _In_ CXPLAT_EVENTQ* queue,
-    _Out_ CXPLAT_CQE* events,
-    _In_ uint32_t count,
-    _In_ uint32_t wait_time // milliseconds
+    _In_ CXPLAT_EVENTQ* Queue,
+    _Out_ CXPLAT_CQE* Events,
+    _In_ uint32_t Count,
+    _In_ uint32_t WaitTime // milliseconds
     )
 {
-    int result = io_uring_peek_batch_cqe(queue, events, count);
-    if (result > 0 || wait_time == 0) return result;
-    if (wait_time != UINT32_MAX) {
+#if DEBUG
+    CxPlatLockAcquire(&Queue->Lock);
+    CXPLAT_DBG_ASSERT(Queue->ContentionCount++ == 0);
+    CxPlatLockRelease(&Queue->Lock);
+#endif
+    int result = io_uring_peek_batch_cqe(&Queue->Ring, Events, Count);
+    if (result > 0 || WaitTime == 0) return result;
+    if (WaitTime != UINT32_MAX) {
         struct __kernel_timespec timeout;
-        timeout.tv_sec = (wait_time / 1000);
-        timeout.tv_nsec = ((wait_time % 1000) * 1000000);
-        (void)io_uring_wait_cqe_timeout(queue, events, &timeout);
+        timeout.tv_sec = (WaitTime / 1000);
+        timeout.tv_nsec = ((WaitTime % 1000) * 1000000);
+        (void)io_uring_wait_cqe_timeout(&Queue->Ring, Events, &timeout);
     } else {
-        (void)io_uring_wait_cqe(queue, events);
+        (void)io_uring_wait_cqe(&Queue->Ring, Events);
     }
-    return io_uring_peek_batch_cqe(queue, events, count);
+    result = io_uring_peek_batch_cqe(&Queue->Ring, Events, Count);
+#if DEBUG
+    CxPlatLockAcquire(&Queue->Lock);
+    CXPLAT_DBG_ASSERT(--Queue->ContentionCount == 0);
+    CxPlatLockRelease(&Queue->Lock);
+#endif
+    return result;
 }
 
 inline
 void
 CxPlatEventQReturn(
-    _In_ CXPLAT_EVENTQ* queue,
-    _In_ uint32_t count
+    _In_ CXPLAT_EVENTQ* Queue,
+    _In_ uint32_t Count
+    )
+{
+#if DEBUG
+    CxPlatLockAcquire(&Queue->Lock);
+    CXPLAT_DBG_ASSERT(Queue->ContentionCount++ == 0);
+    CxPlatLockRelease(&Queue->Lock);
+#endif
+    io_uring_cq_advance(&Queue->Ring, Count);
+#if DEBUG
+    CxPlatLockAcquire(&Queue->Lock);
+    CXPLAT_DBG_ASSERT(--Queue->ContentionCount == 0);
+    CxPlatLockRelease(&Queue->Lock);
+#endif
+}
+
+inline
+CXPLAT_SQE*
+CxPlatCqeGetSqe(
+    _In_ const CXPLAT_CQE* cqe
     )
 {
-    io_uring_cq_advance(queue, count);
+    return (CXPLAT_SQE*)(uintptr_t)(*cqe)->user_data;
+}
+
+inline
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+CxPlatSqeClassicCompletion(
```

Comment:
Why add this interface to just do what the caller was doing anyways?

Reply:
See the other comment (The new completion function supports batching completions) and the comment in this function (it's a hack to avoid updating all code).